src/main/scala/org/ergoplatform/reemission/ReemissionRules.scala

```diff
@@ -0,0 +1,123 @@
+package org.ergoplatform.reemission
+
+import org.ergoplatform.ErgoLikeContext.Height
+import org.ergoplatform.contracts.ReemissionContracts
+import org.ergoplatform.mining.emission.EmissionRules
+import org.ergoplatform.{ErgoAddressEncoder, Pay2SAddress}
+import org.ergoplatform.settings.{ErgoSettings, ReemissionSettings}
+import sigmastate.SBoolean
+import sigmastate.Values.Value
+import sigmastate.eval.CompiletimeIRContext
+import sigmastate.lang.{CompilerSettings, SigmaCompiler, TransformingSigmaBuilder}
+
+import scala.util.Try
+
+/**
+  * Contains re-emission contracts (defined in `ReemissionContracts`) and helper functions
+  * for re-emission.
+  */
+class ReemissionRules(reemissionSettings: ReemissionSettings) extends ReemissionContracts {
+
+  override val reemissionNftIdBytes: Array[Byte] = reemissionSettings.reemissionNftIdBytes
+  override val reemissionStartHeight: Height = reemissionSettings.reemissionStartHeight
+
+  /**
+    * How many ERG taken from emission to re-emission initially
+    */
+  val basicChargeAmount = 12 // in ERG
+
+  /**
+    * @return how many re-emission tokens can be unlocked at given height
+    */
+  def reemissionForHeight(height: Height,
+                          emissionRules: EmissionRules): Long = {
+    val emission = emissionRules.emissionAtHeight(height)
+    if (height >= reemissionSettings.activationHeight &&
+      emission >= (basicChargeAmount + 3) * EmissionRules.CoinsInOneErgo) {
+      basicChargeAmount * EmissionRules.CoinsInOneErgo
+    } else if (height >= reemissionSettings.activationHeight &&
+      emission > 3 * EmissionRules.CoinsInOneErgo) {
+      emission - 3 * EmissionRules.CoinsInOneErgo
+    } else {
+      0L
+    }
+  }
+}
+
+object ReemissionRules {
+
+  /**
+    * @param mainnet - whether to create address for mainnet or testnet
+    * @return - P2S address for a box used to carry emission NFT and re-emission tokens
+    *           to inject them into the emission box on activation height
+    */
+  def injectionBoxP2SAddress(mainnet: Boolean): Pay2SAddress = {
+    val networkPrefix = if (mainnet) {
+      ErgoAddressEncoder.MainnetNetworkPrefix
+    } else {
+      ErgoAddressEncoder.TestnetNetworkPrefix
+    }
+
+    implicit val addrEncoder = new ErgoAddressEncoder(networkPrefix)
+
+    val source =
+      """
+        |  {
+        |    INPUTS(0).value > 30000000L * 1000000000L
```

Comment:
This seems to be a huge number, is that legit?

Reply:
Yes, this is 30M ERGs, see description in EIP-27